### ISSUE TITLE
Photometric reconstruction loss

### ISSUE THREAD

**Issue by hannesdem** [ClementPinard/SfmLearner-Pytorch#86]
Hi,

First of all thank you for implementing this in pytorch. 

I have a question about the calculation of the photometric reconstruction loss. In the file "loss_functions.py" on line 32, there is the following line of code:

`diff = (tgt_img_scaled - ref_img_warped) * valid_points.unsqueeze(1).float()`

Is there a particular reason why you multiply the difference of target and warped image with the valid_points map? How I see it, by doing this, the out of range pixels are set to zero in the difference image. Therefore, these pixels will add a value of zero when calculating the mean over the difference image. Could this result in the network predicting as much out of range pixels as possible because the photometric error becomes zero in that case?

I checked with the tensorflow implementation of the authors and there, they just calculate difference as (target-warped).

Thanks for your answer in advance
Kind regards
Hannes

**Comment by ClementPinard** [comment]
`valid_points` is constructed from a binary tensor, so it's not differentiable (more precisely, the gradient is always 0). As such, this strictly equivalent as pnly selecting valid points to compute the loss.

This is important as for now, `grid_sample` behaves dangerously at boundaries : when a pirxel is between H and H+1 (or any other boundary), the outputted value is an interpolation between gray (color value to exactly 0) and color, and this is not what we want since mix of colored and gray will be incentived to not go gray even if the have to.

Hope it was clear for you,

Clément

**Comment by hannesdem** [comment]
Hi,

Yes it's clear now, thanks for the answer.

Kind regards 
Hannes Demuynck
